File: Python3/1 - 200/143 Reorder List.py
# 84ms, 90.83%; 20.8MB, 100.00%
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reorderList(self, head: ListNode) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if not head: return
        slow, fast = head, head.next
        while fast:
            slow = slow.next
            fast = fast.next
            if fast: fast = fast.next
        link1, link2, slow.next = head, slow.next, None
        st = [node for node in self.__helper(link2)]
        while st:
            node = st.pop()
            link1.next, link1, node.next = node, link1.next, link1.next

    def __helper(self, head: ListNode) -> ListNode:
        while head:
            yield head
            head = head.next

File: 1 - 200/143 Reorder List.cc
// 48ms, 80.40%; 12.6MB, 29.41%
#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    void reorderList(ListNode *head) {
        if (!head) return;
        ListNode *slow = head, *fast = head->next;
        while (fast) { // divide link into 2 parts
            slow = slow->next;
            fast = fast->next;
            if (fast) fast = fast->next;
        }
        ListNode *link1 = head, *link2 = slow->next;
        slow->next = nullptr;
        stack<ListNode *> st;
        while (link2) { // reverse order of link2
            st.push(link2);
            link2 = link2->next;
        }
        while (!st.empty()) { // combine 2 links
            ListNode *tmp = link1->next;
            link1->next = st.top();
            link1->next->next = tmp;
            link1 = tmp;
            st.pop();
        }
    }
};
